core/consensus/pow/src/lib.rs

```diff
@@ -0,0 +1,301 @@
+		justification: Option<Justification>,
+		mut body: Option<Vec<B::Extrinsic>>,
+	) -> Result<(BlockImportParams<B>, Option<Vec<(CacheKeyId, Vec<u8>)>>), String> {
+		let inherent_data = self.inherent_data_providers
+			.create_inherent_data().map_err(String::from)?;
+		let timestamp_now = inherent_data.timestamp_inherent_data().map_err(String::from)?;
+
+		let best_hash = self.client.info().best_hash;
+		let hash = header.hash();
+		let parent_hash = *header.parent_hash();
+		let parent_aux_key = POW_AUX_PREFIX.iter().chain(&parent_hash[..])
+			.cloned().collect::<Vec<_>>();
+		let best_aux_key = POW_AUX_PREFIX.iter().chain(&best_hash[..])
+			.cloned().collect::<Vec<_>>();
+		let best_aux = match self.client.get_aux(&best_aux_key)
+			.map_err(|e| format!("{:?}", e))?
+		{
+			Some(bytes) => PowAux::decode(&mut &bytes[..]).map_err(|e| format!("{:?}", e))?,
+			None => Default::default(),
+		};
+		let mut aux = match self.client.get_aux(&parent_aux_key)
+			.map_err(|e| format!("{:?}", e))?
+		{
+			Some(bytes) => PowAux::decode(&mut &bytes[..]).map_err(|e| format!("{:?}", e))?,
+			None => Default::default(),
+		};
+
+		let (checked_header, difficulty, seal) = self.check_header::<B>(
+			header,
+			BlockId::Hash(parent_hash),
+		)?;
+		aux.total_difficulty += difficulty;
+
+		if let Some(inner_body) = body.take() {
+			let block = B::new(checked_header.clone(), inner_body);
+
+			self.check_inherents(
+				block.clone(),
+				BlockId::Hash(parent_hash),
+				inherent_data,
+				timestamp_now
+			)?;
+
+			let (_, inner_body) = block.deconstruct();
+			body = Some(inner_body);
+		}
+		let key = POW_AUX_PREFIX.iter().chain(&hash[..]).cloned().collect::<Vec<_>>();
+		let import_block = BlockImportParams {
+			origin,
+			header: checked_header,
+			post_digests: vec![seal],
+			body,
+			finalized: false,
+			justification,
+			auxiliary: vec![(key, Some(aux.encode()))],
+			fork_choice: ForkChoiceStrategy::Custom(aux.total_difficulty > best_aux.total_difficulty),
+		};
+
+		Ok((import_block, None))
+	}
+}
+
+/// The PoW import queue type.
+pub type PowImportQueue<B> = BasicQueue<B>;
+
+pub fn import_queue<B, C>(
+	block_import: BoxBlockImport<B>,
+	client: Arc<C>,
+	inherent_data_providers: InherentDataProviders,
+) -> Result<PowImportQueue<B>, consensus_common::Error> where
+	B: BlockT<Hash=H256>,
+	C: 'static + ProvideRuntimeApi + HeaderBackend<B> + BlockOf + ProvideCache<B> + Send + Sync + AuxStore,
+	C::Api: BlockBuilderApi<B> + PowApi<B>,
+{
+	if !inherent_data_providers.has_provider(&srml_timestamp::INHERENT_IDENTIFIER) {
+		match inherent_data_providers.register_provider(srml_timestamp::InherentDataProvider) {
+			Ok(()) => (),
+			Err(e) => warn!("Registering inherent data provider for timestamp failed"),
+		}
+	}
+
+	let verifier = PowVerifier {
+		client: client.clone(),
+		inherent_data_providers,
+	};
+
+	Ok(BasicQueue::new(
+		verifier,
+		block_import,
+		None,
+		None
+	))
+}
+
+pub fn start_mine<B: BlockT<Hash=H256>, I, C, E>(
```

Comment:
That should return a `Future`. Almost all of Substrate uses futures and not background threads.
You can do that by doing something like `futures_timer::Interval::new().for_each(|_| mine_one())`.

Reply:
The issue is, mine_one is not a yieldable operation at all. Any thread running it basically causes that thread to spin at 100% for all time, and have no chance to process anything else. Putting this under a future, I think, will pollute the future pool. It may *appear* fine in a threadpool future runtime, but if the future runtime is single threaded, it breaks all other running futures.

I do think we can put the initial block builder under a future, but when doing the actual mining, I still think we need to use a background thread.

Reply:
If we decided to leave this as is, we should leave a note in the comment on saying that this is rather an exception, so people won't take this as an example way of doing this (or at least they understood the tradeoffs)

Reply:
If runtime calls ever become asynchronous, we can make `mine_one` yieldable. But right now, you're correct that a separate thread might be a good idea.

Reply:
@DemiMarie-parity the algorithm is defined as a trait and you can implement whatever you want, be it CPU-friendly, GPU-friendly or ASIC-friendly.